app/src/main/java/io/appium/uiautomator2/handler/W3CActions.java

```diff
@@ -0,0 +1,276 @@
+     * <li>size - A normalized value that describes the approximate size of the pointer touch area
+     * in relation to the maximum detectable size of the device.
+     * It represents some approximation of the area of the screen being
+     * pressed; the actual value in pixels corresponding to the
+     * touch is normalized with the device specific range of values
+     * and scaled to a value between 0 and 1 (the default value).
+     * The value of size can be used to determine fat touch events.</li>
+     * </ul>
+     * <p>
+     * Applicable key and meta key codes for <b>key</b> input source can be found in
+     * {@link KeyEvent} documentation. Value transformation to a numeric key code is
+     * done via {@link String#charAt(int)} method call, which means, for example,
+     * that the value <em>"\\u2000"</em> equals to meta key code 0x2000
+     * {@link KeyEvent#META_CTRL_LEFT_ON}. Meta key codes are also applied to pointer
+     * events, which are happening at the same moment.
+     * <p>
+     *
+     * @param request JSON request formatted according to W3C actions endpoint compilation rules.
+     * @return The standard {@link AppiumResponse} instance with return value or error code inside.
+     */
+    @Override
+    public AppiumResponse safeHandle(IHttpRequest request) {
+        try {
+            final JSONArray actions = ActionsHelpers.preprocessActions(
+                    (JSONArray) getPayload(request).get("actions")
+            );
+
+            if (API.API_18) {
+                if (executeActions(actions)) {
+                    return new AppiumResponse(getSessionId(request), WDStatus.SUCCESS, "OK");
+                }
+                return new AppiumResponse(getSessionId(request), WDStatus.UNKNOWN_ERROR,
+                        "Unable to perform W3C actions. Check the logcat output " +
+                                "for possible error reports and make sure your input actions chain is valid.");
+            }
+            Logger.error("Device does not support API < 18!");
+            return new AppiumResponse(getSessionId(request), WDStatus.UNKNOWN_ERROR,
+                    "Cannot perform W3C actions on device below API level 18");
+        } catch (JSONException | ActionsParseException e) {
+            Logger.error("Exception while reading JSON: ", e);
+            return new AppiumResponse(getSessionId(request), WDStatus.JSON_DECODER_ERROR, e);
+        } catch (Exception e) {
+            Logger.error("Exception while performing W3C action: ", e);
+            return new AppiumResponse(getSessionId(request), WDStatus.UNKNOWN_ERROR, e);
+        }
+    }
+
+    private boolean injectEventSync(InputEvent event) {
+        return UiAutomatorBridge.getInstance().getInteractionController().injectEventSync(event);
+    }
+
+    private static final List<Integer> HOVERING_ACTIONS = Arrays.asList(
+            MotionEvent.ACTION_HOVER_ENTER, MotionEvent.ACTION_HOVER_EXIT, MotionEvent.ACTION_HOVER_MOVE
```

Comment:
Indentation?

Reply:
I just use the default Android Studio formatter